В цьому домашньому завданні ми реалізуємо логістичну регресію на `numpy`.
Ці завдання допоможуть вам ґрунтовно засвоїти основні концепції логістичної регресії та реалізувати їх на практиці 🔥

#### Завдання 1: Реалізація функції сигмоїди
1. З використанням `numpy` напишіть функцію `sigmoid(z)` для обчислення значення сигмоїди згідно з формулою:
   $$
   \sigma(z) = \frac{1}{1 + e^{-z}}
   $$
2. Використовуючи цю функцію, обчисліть значення сигмоїди для наступних даних: $ z = [-2, -1, 0, 1, 2] $. Виведіть результат обчислень.


In [ ]:
#1
import numpy as np

def sigmoid(z):
    return 1 / (1 + np.exp(-z))

In [ ]:
#2
z = np.array([-2, -1, 0, 1, 2])

sigmoid_values = sigmoid(z)

print("Sigmoid values for z =", z, ":", sigmoid_values.round(5))

Sigmoid values for z = [-2 -1  0  1  2] : [0.1192  0.26894 0.5     0.73106 0.8808 ]


Функція sigmoid(z) обчислює значення сигмоїди для заданого значення z.

Як і передбачалося, обчислення сигмоїдної функції для наших даних дало результати, які знаходяться в межах від 0 до 1.

Значення функції для z = [-2 -1  0  1  2] становлять: [0.1192  0.26894 0.5     0.73106 0.8808 ] відповідно.




#### Завдання 2: Реалізація функції гіпотези для логістичної регресії
1. Напишіть функцію `hypothesis(theta, X)`, яка обчислює гіпотезу для логістичної регресії, використовуючи функцію сигмоїди. Формула гіпотези:
   $$
   h_\theta(x) = \sigma(\theta^T x) = \frac{1}{1 + e^{-\theta^T x}}
   $$
2. Використайте функцію `hypothesis` для обчислення значень гіпотези для наступних даних:
   
   $\theta = [0.5, -0.5]$
   
   $X = \begin{bmatrix} 1 & 2 \\ 1 & -1 \\ 1 & 0 \\ 1 & 1 \end{bmatrix}$

  Виведіть результат обчислень.


In [ ]:
def hypothesis(theta, X):
    z = np.dot(X, theta)
    return sigmoid(z)


In [ ]:
# Задані дані
theta = np.array([0.5, -0.5])
X = np.array([[1, 2], [1, -1], [1, 0], [1, 1]])

# Викликаємо функцію гіпотези для наших значень
hypothesis_values = hypothesis(theta, X)

In [ ]:
# Виводимо результати обчислень
print("Hypothesis values:", hypothesis_values)


Hypothesis values: [0.37754067 0.73105858 0.62245933 0.5       ]


Кожне значення в масиві, яке повертається функцією hypothesis, є передбаченою ймовірністю того, що мітка для відповідного прикладу дорівнює 1.


Чим ближче значення до 1, тим більше модель впевнена, що результат для цього прикладу є позитивним y=1, чим ближче значення до 0, тим більше модель схиляється до того, що результат для цього прикладу є негативним мітка y=0.

#### Завдання 3: Реалізація функції для підрахунку градієнтів фукнції втрат
1. Напишіть функцію `compute_gradient(theta, X, y)`, яка обчислює градієнти функції втрат для логістичної регресії. Формула для обчислення градієнта:
   $$
   \frac{\partial L(\theta)}{\partial \theta_j} = \frac{1}{m} \sum_{i=1}^{m} \left[ (h_\theta(x^{(i)}) - y^{(i)}) x_j^{(i)} \right]
   $$
2. Використайте функцію `compute_gradient` для обчислення градієнтів для наступних даних:

  $\theta = [0.5, -0.5]$

  $X = \begin{bmatrix} 1 & 2 \\ 1 & -1 \\ 1 & 0 \\ 1 & 1 \end{bmatrix}$

  $y = [1, 0, 1, 0]$

  Виведіть результат обчислень.

In [ ]:
import numpy as np

def compute_gradient(theta, X, y):
    m = X.shape[0]  # Number of training examples
    h0 = hypothesis(theta, X)  # Hypothesis values
    error = h0 - y
    gradient = (1 / m) * np.dot(X.T, error)
    return gradient


In [ ]:
theta = np.array([0.5, -0.5])

X = np.array([
    [1, 2],
    [1, -1],
    [1, 0],
    [1, 1]
])

# True labels
y = np.array([1, 0, 1, 0])

In [ ]:
# Обчислення градієнтів для наших даних
gradient = compute_gradient(theta, X, y)

# Виводимо результати обчислень
print("Gradient values:", gradient)

Gradient values: [ 0.05776464 -0.36899431]


Наша функція повернула значення першого елемента градієнта, що відповідає частковій похідній функції втрат по theta0 і становить 0.05776464.

Другий елемент градієнта відповідає частковій похідній функції втрат по theta1 дорівнює 0.36899431.



#### Завдання 4: Реалізація повного батч градієнтного спуску

**Задача:**
1. Напишіть функцію `full_batch_gradient_descent(X, y, lr=0.1, epochs=100)`, яка реалізує алгоритм Full градієнтного спуску для логістичної регресії. Використовуйте такі формули:
   - Гіпотеза: $ h_\theta(x) = \sigma(\theta^T x) $
   - Оновлення параметрів: $ \theta_j := \theta_j - \alpha \frac{\partial L(\theta)}{\partial \theta_j} $
2. Використайте функцію `full_batch_gradient_descent` для обчислення параметрів моделі на наступних даних:

  $X = \begin{bmatrix} 1 & 2 \\ 1 & -1 \\ 1 & 0 \\ 1 & 1 \end{bmatrix}$

  $y = [1, 0, 1, 0]$

  Увага! Матриця $X$ вже має стовпець одиниць і передбачається, що це. - стовпець для intercept - параметра зсуву.

  Виведіть результат обчислень.


In [ ]:
def full_batch_gradient_descent(X, y, lr=0.1, epochs=100):
    X = np.array(X)
    y = np.array(y)
    theta = np.zeros(X.shape[1])  # Ініціалізація параметрів нулями

    for epoch in range(epochs):
        gradient = compute_gradient(theta, X, y)  # Обчислення градієнта
        theta -= lr * gradient  # Оновлення параметрів

    return theta


In [ ]:
# Дані
X = [
    [1, 2],
    [1, -1],
    [1, 0],
    [1, 1]
]
y = [1, 0, 1, 0]

# Навчаємо модель за допомогою градієнтного спуску
theta = full_batch_gradient_descent(X, y, lr=0.1, epochs=100)
print("Параметри моделі:", theta)

Параметри моделі: [-0.2893693   0.77655125]


Згідно наших обчислень:

Інтерсепт (theta0) становить -0.2893693.
Коефіцієнт theta1  становить 0.77655125.


#### Завдання 5. Обчислення точності моделі

1. Напишіть функцію `predict_proba(theta, X)`, яка використовує знайдені параметри $\theta$ для обчислення ймовірностей належності поточного прикладу з даних до класу $y=1$ на основі значень $\sigma(\theta^T x)$.

2. Напишіть функцію `predict(theta, X, threshold=0.5)`, яка обчислює клас з передбаченої імовірності належності екземпляра до класу 1 з порогом 0.5. Тобто якщо ймовірність менше 0.5, то передбачаємо клас 0, інакше клас 1.

3. Напишіть функцію `accuracy(y_true, y_pred)`, яка обчислює точність моделі, визначивши частку правильно передбачених класів.

  Формула метрики Accuracy:
  $$
  \text{Accuracy} = \frac{\sum_{i=1}^{m} I(\hat{{y}^{(i)}} = y^{(i)})}{m}
  $$

  де $\hat{{y}^{(i)}}$ - передбачене значення класу, $I$ - індикаторна функція (яка дорівнює 1, якщо умова виконується, і 0 - якщо ні), $m$ - кількість прикладів.

4. Обчисліть з використанням даних в завданні 4 $X$, $y$ та обчислених коефіцієнтах $\theta$ та виведіть на екран:
  - передбачені моделлю імовірності належності кожного з екземплярів в матриці `X` до класу 1
  - класи кожного екземпляра з матриці `X`
  - точність моделі.

In [ ]:
import numpy as np

def sigmoid(z):
    return 1 / (1 + np.exp(-z))

def predict_proba(theta, X):
    z = np.dot(X, theta)
    return sigmoid(z)

def predict(theta, X, threshold=0.5):
    probabilities = predict_proba(theta, X)
    return (probabilities >= threshold).astype(int)

In [ ]:
def accuracy(y_true, y_pred):

    # Обчислюємо кількість правильних передбачень
    correct_predictions = np.sum(y_true == y_pred)
    # Загальна кількість прикладів
    total_samples = len(y_true)
    # Точність моделі
    return correct_predictions / total_samples

In [ ]:
# Дані
X = np.array([
    [1, 2],
    [1, -1],
    [1, 0],
    [1, 1]
])
y = np.array([1, 0, 1, 0])

theta = np.array([-0.2893693, 0.77655125])

In [ ]:
# Обчислення ймовірностей, передбачених класів і точності
probabilities = predict_proba(theta, X)
predicted_classes = predict(theta, X)
model_accuracy = accuracy(y, predicted_classes)

# Виведення результатів
print("Ймовірності належності кожного екземпляра до класу 1:")
print(probabilities)

print("Передбачені класи для кожного екземпляра:")
print(predicted_classes)

print("Точність моделі:")
print(model_accuracy)

Ймовірності належності кожного екземпляра до класу 1:
[0.77966809 0.25617966 0.42815828 0.61944235]
Передбачені класи для кожного екземпляра:
[1 0 0 1]
Точність моделі:
0.5


Модель правильно класифікувала лише 2 з 4 прикладів, що дає точність 0.5 (50%).